### ISSUE TITLE
Request - GPX Build in

### ISSUE THREAD

**Issue by NoobWithBoobs** [jabbink/PokemonGoBot#977]
Hello,
Requesting for GPX following to be build in to json/config file. Keep the box from traveling off course.

Do-able?

**Comment by Peyphour** [comment]
Would be interesting

**Comment by NoobWithBoobs** [comment]
and gpx are pretty easy to make with google earth and bunch of other programs too.

**Comment by Peyphour** [comment]
Support for GPX was already in the previous version of the GUI so it shouldn't be too hard to implement

**Comment by NoobWithBoobs** [comment]
I have the bot reload every 30mins due and usually i'm not around to resubmit via the gui again and again. If gpx support was build into config/json file, bot could roam around in more focus areas.

**Comment by Peyphour** [comment]
Just checked, it's do-able

**Comment by NoobWithBoobs** [comment]
The excitement rages!!!

**Comment by staurakis** [comment]
It would be a great add on!!! :)

**Comment by Peyphour** [comment]
Will do it today or tomorrow

**Comment by NoobWithBoobs** [comment]
i look forward to it, keep us posted here 👍

**Comment by Peyphour** [comment]
Will have two settings I think : 
* GPX file
* Number of time the file must be repeated

**Comment by NoobWithBoobs** [comment]
Repeat is an awesome idea, please be sure to include -1 option as well.

**Comment by Peyphour** [comment]
Of course ^^

**Comment by smileytechguy** [comment]
How can I make this kind of GPX?

**Comment by NoobWithBoobs** [comment]
Easist ways to make GPX files online
http://www.gpsies.com/createTrack.do?language=en

**Comment by NoobWithBoobs** [comment]
@Peyphour Direction might be cool in future versions.

**Comment by Peyphour** [comment]
Directions ?

**Comment by NoobWithBoobs** [comment]
Direction as in backwards, forward, bidirectional.

**Comment by staurakis** [comment]
I think the only logical route to have is a circular one... There is no point to directly go back the opposite route as nothing would have been spawn yet.
On the other hand, a delay time maybe a good idea if you have a small route ;)

**Comment by NoobWithBoobs** [comment]
@staurakis, you do have a point with circular routes but I was thinking long roads and such. Anyways, I'm excited for just the basic, I could make a loop myself. 

Delay Time to wait for Pokemon to spawn?

**Comment by staurakis** [comment]
Me too! ;)

**Comment by Peyphour** [comment]
#1008

**Comment by NoobWithBoobs** [comment]
damn exciting, can't wait to see it be included in the main build!!

**Comment by Peyphour** [comment]
Can you test it ? I can't do it right now ^^

**Comment by NoobWithBoobs** [comment]
yep, pulling as I type!

**Comment by NoobWithBoobs** [comment]
ink.abb.pogo.scraper.TestSettings > testDefaultSettings FAILED
    java.io.FileNotFoundException at TestSettings.kt:25

ink.abb.pogo.scraper.TestSettings > testTemplateKeys FAILED
    java.io.FileNotFoundException at TestSettings.kt:73

2 tests completed, 2 failed
:test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///Users/Katherine/Desktop/Beta/PokemonGoBot/build/reports/tests/index.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 10 mins 37.845 secs

**Comment by Peyphour** [comment]
10min for building ?!

Are you sure you have a config.properties.template ? Build passed on Travis so this isn't my fault

**Comment by NoobWithBoobs** [comment]
mhmm redo time!

**Comment by NoobWithBoobs** [comment]
So I've tested it now for 20mins and I don't see it following the gpx patch. Could you provide your gpx file and which one is correct?

```
# GPX file to be added on the bot's start
gpx_file=/Users/Katherine/Desktop/Beta/PokemonGoBot/WalkInCentralPark.gpx
# Number of times the bot must repeat this GPX file
gpx_repeat=-1
```

or 

# GPX file to be added on the bot's start
gpx_file=WalkInCentralPark.gpx
# Number of times the bot must repeat this GPX file
gpx_repeat=-1

**Comment by Peyphour** [comment]
If gpxRepeat is -1 or 0, the bot will not follow the path at all. Literally : number of times the bot will repeat the path, so at least 1

**Comment by Peyphour** [comment]
And the path is relative to the jar file

**Comment by Peyphour** [comment]
Well the formulation may be wrong in this one

**Comment by NoobWithBoobs** [comment]
I have to run outside but i'll come back and test some more in two-three hours. Still very excited 👍

**Comment by Peyphour** [comment]
Going to sleep, we will continue tomorrow

**Comment by NoobWithBoobs** [comment]
gpx file loads but it's still going to latitude/longitude settings in config file. Am I suppose to remove those?

**Comment by Peyphour** [comment]
What does it do after going to config coords ? Could you provide a bit of logs (via pastebin)?

**Comment by NoobWithBoobs** [comment]
@Peyphour currently dealing with soft ban, i'm sure you noticed it too

**Comment by Peyphour** [comment]
Nope not banned ^^ I think I know what the problem is, will keep you posted

**Comment by NoobWithBoobs** [comment]
I hope it's not hard ban! Please keep me posted, i'd be very grateful. 👍

**Comment by staurakis** [comment]
@NoobWithBoobs i'm willing to help by any means (and that's obvious i think) but i'm a bit out of my expertise here... If i receive some rough guidelines, i can and will be more than happy to do so! :)

**Comment by Peyphour** [comment]
Just run the bot with a GPX file, let it run for 4-5 minutes et send me : the GPX file and the logs

**Comment by staurakis** [comment]
i have the compiled version @Peyphour .... As far as i understand, i have to go for the other way to run it and get the files you have changed.... That is the part i am missing.... That's the part that i need a bit of info. I am missing the platform to 'compile' it.... I think at least.....

**Comment by Peyphour** [comment]
Forget it then, I'll do it myself tomorrow

**Comment by staurakis** [comment]
i think i did it.... building it right now....
I will test it with my gpx and let you know ;)

**Comment by staurakis** [comment]
fixed it...continuing.....

**Comment by Peyphour** [comment]
Sorry, didn't see your issue, you need JDK 1.8

**Comment by staurakis** [comment]
yeap yeap figured it out ;)
checking the gpx right now ;)

**Comment by staurakis** [comment]
@Peyphour my gpx format is a bit different from your test.gpx.
Also, don't you have any 'switch' (as you describe in your code) in the configuration file? I haven't seen anything.

**Comment by staurakis** [comment]
I have added two variables in to the configuration file: gpxFile and gpxRepeat in a try to make it read my gpx file but without any luck.
I have also changed my gpx xml file in accordance to your xml but again it didn't help.
@Peyphour Could you please help on how to make it "read" the gpx file?

**Comment by Peyphour** [comment]
What do you mean by 'switch' ? 
Could you provide via pastebin your config file and GPX file ?

**Comment by staurakis** [comment]
i think i found it.... you name your variables with _ in between. I had them in the configuration file as you describe them in the comments without anything.
Testing again.

**Comment by staurakis** [comment]
I can get it "read" the gpx file.... :(
@Peyphour Am i doing something wrong....?

**Comment by Peyphour** [comment]
Can't help you at the moment, it's 11 PM here in Paris, we can talk about this tomorrow

**Comment by staurakis** [comment]
ok. whenever you can

**Comment by Peyphour** [comment]
Hello, any updates ? Do you still need help ?

**Comment by staurakis** [comment]
Yes i need help on this. I couldn't make it 'read' the gpx file. Basicaly it didn't do anything.
Although currently i have paused testing due to ban issue, i will begin again in few hours. So if tou please help on this...

**Comment by NoobWithBoobs** [comment]
Be care, I'm ban. I propose catch, stop, and time limit to avoid getting banned.

**Comment by AlexBoulyga** [comment]
I have added those 3 limits, however they currently simply stop (not kill) the bot, without restarting it later vor anything like that
In #1047

**Comment by NoobWithBoobs** [comment]
@AlexBoulyga, very nice. Checking it out

**Comment by AlexBoulyga** [comment]
If you find any bugs in that functionality (try it out with low values) Please tell me. However, AS a Warning, it currently stops the bot but the process seems to still run, so you need to kill it manually (idk why)

**Comment by Peyphour** [comment]
A little off topic here guys :)
@staurakis, what did you try to make it work ? The options are 'gpxFile' (the path MUST be relative to the jar file/project if you build from source), and 'gpxRepeat' (any value >0 works, 1 should be good for testing)

**Comment by NoobWithBoobs** [comment]
@peyphour, you're right. @AlexBoulyga I'll report back as soon as I get a chance. Currently at work.


@peyphour, considering the recent bans that we've all been impacted by. Would it possible if the bot didn't follow the gpx perfectly. Meaning the path stays the same but the coordinates numbers have a bit of change. So it seems less like a bot and more like a drunk/teenager person walking.

**Comment by Peyphour** [comment]
Yeah there's already a part of the code which does some randomization, I will adapt it

**Comment by NoobWithBoobs** [comment]
@Peyphour Genius!

**Comment by Peyphour** [comment]
Actually, I just re-checked and the randomization of the path is done for every coordinate supplied to the bot (including when going to pokestop, catching pokemons etc ...)

**Comment by Peyphour** [comment]
I did a test and it works for me, double check your settings

**Comment by NoobWithBoobs** [comment]
By chance, would you share your settings with us? Maybe we're missing something

**Comment by Peyphour** [comment]
Settings : http://pastebin.com/FeRP4R2r
GPX file : https://raw.githubusercontent.com/Peyphour/PokemonGoBot/ce0b1422b1388d9bcfc2fe9b22763731fb4e6e29/test.gpx
